pkg/kubectl/apply/element.go

```diff
@@ -0,0 +1,331 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+import (
+	"fmt"
+)
+
+// FieldMeta defines the strategy used to apply a Patch for an element
+type FieldMeta interface {
+	// MergeType specifies how a field should be merged.  One of
+	// replace, merge, replace-keys
+	GetFieldMergeType() string
+	// MergeKey specifies which fields to use in order to merge items
+	// in lists
+	GetFieldMergeKey() []string
+
+	GetFieldType() string
+}
+
+type FieldMetaImpl struct {
+	// The type of merge strategy to use for this field
+	// maybe "merge", "replace" or "retainkeys"
+	MergeType string
+	// The merge key to use when the MergeType is "merge" and underlying type is a list
+	MergeKey []string
+	// The openapi type of the field - "list", "primitive", "map"
+	Type string
+}
+
+func (s FieldMetaImpl) GetFieldMergeType() string {
+	return s.MergeType
+}
+
+func (s FieldMetaImpl) GetFieldMergeKey() []string {
+	return s.MergeKey
+}
+
+func (s FieldMetaImpl) GetFieldType() string {
+	return s.Type
+}
+
+// Element contains the record, local, and remote value for a field in an object
+// as well as the merge strategy to use
+type Element interface {
+	// FieldMeta specifies which merge strategy to use for this element
+	FieldMeta
+
+	// Accept calls the type specific method on Visitor passing in itself
+	// e.g. Accept on a ListElement will call Visitor.VisitList(self)
+	// Returns the Result of merging the elements using the passed in
+	// strategy
```

Comment:
I am good with that.  WDYT of the Visitor interface being `MergeStrategy` and the Accept method being `Execute`?

Reply:
I think I'm fine giving up on the typical visitor naming convention, especially since the semantic (mostly because of the return type) is  focusing on merge. Let's comment explicitly that it is using a visitor pattern though.

Since it's merging, should it be named `Merge`?

`Elements.Merge(MergeStrategy)` looks *very* clear to me!

Reply:
Makes sense to me.  I am going to sleep on this.  One weird thing is we can have a "replace" merge strategy or a "merge" merge strategy.  Perhaps there is something better we can come up with.

Reply:
Maybe?

```go
apply.Merge(Strategy) (interface{}, operation, error)
```